src/Interpreters/Cache/FileCache.cpp

```diff
@@ -609,17 +609,19 @@ bool FileCache::tryReserve(FileSegment & file_segment, size_t size)
 
         if (releasable)
         {
-            removed_size += segment_metadata->size();
-            --queue_size;
-
             auto segment = segment_metadata->file_segment;
             if (segment->state() == FileSegment::State::DOWNLOADED)
             {
                 const auto & key = segment->key();
+
                 auto it = to_delete.find(key);
                 if (it == to_delete.end())
                     it = to_delete.emplace(key, locked_key.getKeyMetadata()).first;
                 it->second.add(segment_metadata);
+
+                removed_size += segment_metadata->size();
```

Comment:
AFAIU, removed non-DOWNLOADED segments size were accounted twice. So, the fix for `removed_size` seems correct.

But `queue_size` calculation looks incorrect since actual queue size can change when non-DOWNLOADED elements are removed
It'd be more straightforward to calculate the number of segments to remove and subtract it from the actual queue size in `is_main_priority_overflow` check